presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java

```diff
@@ -0,0 +1,481 @@
+                }
+            }
+
+            return new AggregationNode(
+                    idAllocator.getNextId(),
+                    source,
+                    aggregations.build(),
+                    functions.build(),
+                    Collections.<Symbol, Symbol>emptyMap(),
+                    node.getGroupingSets(),
+                    node.getStep(),
+                    node.getSampleWeight(),
+                    node.getConfidence(),
+                    Optional.<Symbol>empty(),
+                    node.getGroupIdSymbol());
+        }
+
+        @Override
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<AggregateInfo>> context)
+        {
+            Optional<AggregateInfo> aggregateInfo = context.get();
+            // presence of aggregateInfo => mask also present
+            if (aggregateInfo.isPresent() && aggregateInfo.get().getMask().get().equals(node.getMarkerSymbol())) {
+                PlanNode source = context.rewrite(node.getSource(), Optional.empty());
+
+                List<Symbol> allSymbols = new ArrayList<>();
+                List<Symbol> groupBySymbols = aggregateInfo.get().getGroupBySymbols(); // a
+                List<Symbol> nonDistinctAggregateSymbols = aggregateInfo.get().getOriginalNonDistinctAggregateArgs(); //b
+                Symbol distinctSymbol = Iterables.getOnlyElement(aggregateInfo.get().getOriginalDistinctAggregateArgs()); // c
+
+                allSymbols.addAll(groupBySymbols);
+                allSymbols.addAll(nonDistinctAggregateSymbols);
+                allSymbols.add(distinctSymbol);
+
+                // If same symbol present in aggregations on distinct and non-distinct values, e.g. select sum(a), count(distinct a),
+                // then we need to create a duplicate stream for this symbol
+                Symbol duplicatedDistinctSymbol = distinctSymbol;
+
+                if (nonDistinctAggregateSymbols.contains(distinctSymbol)) {
+                    Symbol newSymbol = symbolAllocator.newSymbol(distinctSymbol.getName(), symbolAllocator.getTypes().get(distinctSymbol));
+                    nonDistinctAggregateSymbols.set(nonDistinctAggregateSymbols.indexOf(distinctSymbol), newSymbol);
+                    duplicatedDistinctSymbol = newSymbol;
+
+                    // Add project node to get stream for new symbol
+                    ImmutableMap.Builder builder = new ImmutableMap.Builder();
+                    for (Symbol symbol : node.getSource().getOutputSymbols()) {
+                        builder.put(symbol, symbol.toSymbolReference());
+                    }
+                    // Cast to avoid UnaliasSymbolReference from removing this projection
+                    // TODO: find better way to prevent UnaliaSymbolRefrence optimizing out the project
+                    builder.put(newSymbol, new Cast(distinctSymbol.toSymbolReference(),
```

Comment:
It's a correctness issue, which we treat with very high priority. We'll fix it shortly.